src/methods.rs

```diff
@@ -30,16 +33,37 @@ declare_lint!(pub WRONG_SELF_CONVENTION, Warn,
 declare_lint!(pub WRONG_PUB_SELF_CONVENTION, Allow,
               "defining a public method named with an established prefix (like \"into_\") that takes \
                `self` with the wrong convention");
+declare_lint!(pub CALLING_MAIN, Warn,
+              "calling into main detected");
 
 impl LintPass for MethodsPass {
     fn get_lints(&self) -> LintArray {
         lint_array!(OPTION_UNWRAP_USED, RESULT_UNWRAP_USED, STR_TO_STRING, STRING_TO_STRING,
-                    SHOULD_IMPLEMENT_TRAIT, WRONG_SELF_CONVENTION)
+                    SHOULD_IMPLEMENT_TRAIT, WRONG_SELF_CONVENTION, CALLING_MAIN)
     }
 }
 
 impl LateLintPass for MethodsPass {
     fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        fn no_std_present(cx: &LateContext) -> bool {
+        	let attrs : &[Attribute] = &cx.krate.attrs; // TODO replace these two...
+        	for attr in attrs { // lines with ...
+        	// for &attr in cx.krate.attrs { ... once type inf on nightly fixed
+                if let MetaWord(ref mword) = attr.node.value.node {
+                    if mword == &"no_std" { return true; }
+                }
+            }
+            false
+        }
+        if let ExprCall(ref func, ref args) = expr.node {
+            if args.is_empty() {
+                if let ExprPath(_, path) = func.node.clone() {
+                    if !path.global && path.segments[0].identifier.name.as_str() == "main" && !no_std_present(cx) {
```

Comment:
In the `if let` I'm referencing the enum `expr.node` so clearly that's in scope but `expr.node.id` does not exist, nor does a `get` method on `def_map`. The latter is a refcell with a hashmap so presumably I need to just get at the contents but I haven't yet been able to figure out how to get at it. (Slow going, but at least the merge and conflict resolution seem to have gone ok.)